tools/clang/lib/Sema/SemaDecl.cpp

```diff
@@ -8108,7 +8108,8 @@ Sema::ActOnFunctionDeclarator(Scope *S, Declarator &D, DeclContext *DC,
     AddToScope = false;
   }
 
-  if (getLangOpts().HLSL) {
+  if (getLangOpts().HLSL && D.isFunctionDefinition() && D.hasName() &&
+      NewFD->getDeclContext()->getRedeclContext()->isTranslationUnit()) {
```

Comment:
The only diagnostic that seems to use the name is `DiagnoseNodeEntry` and that only for actual diagnostic errors. I don't know how a function could be missing a name though. What prompted these additions?

Reply:
The additional conditions were necessary to prevent accidentally matching a declaration with the same name that was outside of the global scope. Additionally, the function should only be called on top level decl's that could be the Entry function. I don't think any entry function decl will not have a name, so the condition will act as an optimization. There are also some declarations that lack names that end up appearing inside DiagnoseEntry, so the condition skips over those cases.